[Rachel Angelyn Gunawan] iP

Rachel-AG opened the pull request:
# Duke But Better
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

### Duke offers simple task management that is:
- text-based
- easy to use
- ~fast~ very FAST

### All you need to do is:
1. Download the latest version from [here](https://github.com/Rachel-AG/ip/releases)
2. Double click the jar file
3. Add your tasks, deadlines, and events!
4. Let Duke remember it for you 😄 

The best thing is... **IT'S FREE**

### Features:
- [x] Manage your tasks
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method: 
```
    public static void main(String[] args) {
        new Duke("data/duke.txt").run();
    }
````

Commit:
Add Gradle support

Rachel-AG committed:
implemented greet, echo, exit

Rachel-AG committed:
implemented add, list

Rachel-AG committed:
added Task class

Rachel-AG committed:
implemented mark as done

Rachel-AG committed:
added Deadline, Event, and ToDo class, modified Task class

Rachel-AG committed:
implemented todos, events, and deadlines

Rachel-AG committed:
implemented automated text ui testing, modified some messages

Rachel-AG committed:
implemented error handling

Rachel-AG committed:
fixed expected output due to changes in message

Rachel-AG committed:
implemented delete and modified error handling

Rachel-AG committed:
Add data folder and duke.txt

Rachel-AG committed:
Add Storage class

Rachel-AG committed:
Deadline class: add DeadlineFromData and toFile methods
Event class: add EventFromData and toFile methods
ToDo class: add ToDoFromData and toFile methods
Task class: add toFile method

Rachel-AG committed:
Add a new constructor and addTaskFromStr method

Rachel-AG committed:
Add storage as member of Duke and implement save function

Rachel-AG committed:
Store dates as LocalDate objects

Rachel-AG committed:
Merge branch 'branch-Level-7'

Rachel-AG committed:
Merge branch 'branch-Level-8'

Rachel-AG committed:
Extract Ui, Parser, and Command classes

Rachel-AG committed:
Modify constructor arguments and factory method of Deadline and Event classes

Rachel-AG committed:
TaskList class: improve error handling, more descriptive method name
Storage class: remove unused method
Duke class: add more OOP increment

Rachel-AG committed:
Organize classes into packages: duke, duke.command, duke.task

Rachel-AG committed:
Merge pull request #1 from nus-cs2103-AY2223S1/add-gradle-support

Add Gradle support

Rachel-AG committed:
Merge branch 'master' of https://github.com/Rachel-AG/ip

Rachel-AG committed:
Modify mainClassName in build.gradle

Rachel-AG committed:
TaskList class: remove an unused constructor, addTaskFromDataStr method,
rename toFile method to toStorageString method
Parser class: move Parser class into duke.parser package
DeadlineCommand, DeleteCommand, EventCommand, MarkCommand, ToDoCommand,
UnmarkCommand class: rename toFile to toStorageString

Rachel-AG committed:
Deadline, Event, ToDo class: remove unused factory method, add new
constructor, rename toFile method to toStorageString method
Task class: rename toFile method to toStorageString method
Parser class: fix typo
Ui class: fix typo
Storage class: modify load method to return a TaskList object
Duke class: modify constructor due to changes in Storage class

Rachel-AG committed:
Add JUnit tests for Parser, Deadline, Event, and ToDo class

Rachel-AG committed:
Fix bug that occur when data folder does not exist

Rachel-AG committed:
Add find command

Rachel-AG committed:
Add documentations

Rachel-AG committed:
Merge branch 'branch-Level-9'

Rachel-AG committed:
Fix indentation, imports format, and change class variables to private

Rachel-AG committed:
Merge branch 'branch-A-JavaDoc'

Rachel-AG committed:
Merge branch 'branch-A-CodingStandard'

lizekai-richard commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,16 @@
+import duke.Storage;
+import duke.TaskList;
+import duke.Ui;
+
```
I think adding a header here would be better 🤔

Rachel-AG replied:
Yep, will add more headers later. Thank you for the feedback! :)

lizekai-richard commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,37 @@
+        this.isTerminator = isTerminator;
+    }
+
+    public boolean isTerminatorGetter() {
```
Perhaps your method name should be isTerminator?

Rachel-AG replied:
Yeah, I think I should change my method name using verbs. Maybe something like isTerminator or getIsTerminator.

lizekai-richard reviewed (commented):
I like the way you follow coding standard overall 👍

Rachel-AG committed:
Add basic GUI to duke

Rachel-AG committed:
Improve GUI and fix list command display bug

Rachel-AG committed:
Add checkstyle and fix coding style violations

Rachel-AG committed:
Merge branch 'branch-Level-10'

Ugholaf commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,36 @@
+import duke.storage.Storage;
+import duke.task.TaskList;
+
+/**
+ * The Command class represents the command entered by user.
+ */
+public abstract class Command {
```
I like your use of abstract class here to abstract the different types of commands!

Ugholaf commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,116 @@
+        String[] arg = input.split(" ", 2);
+        String command = arg[0];
+        String commandArg = arg.length == 2 ? arg[1].trim() : "";
+        switch (command) {
+        case "list":
+            if (!commandArg.equals("")) {
+                throw new DukeException("list must not have an argument.");
+            }
+            return new ListCommand();
+        case "mark":
+            return new MarkCommand(parseIntArg(commandArg));
+        case "unmark":
+            return new UnmarkCommand(parseIntArg(commandArg));
+        case "delete":
+            return new DeleteCommand(parseIntArg(commandArg));
+        case "find":
+            return new FindCommand(commandArg);
+        case "todo":
+            if (commandArg.equals("")) {
+                throw new DukeException("A todo must contain a description.");
+            }
+            return new ToDoCommand(new ToDo(commandArg));
+        case "deadline":
+            String[] deadlineArgs = getDeadlineArgs(commandArg);
+            return new DeadlineCommand(new Deadline(deadlineArgs[0], parseDateArg(deadlineArgs[1])));
+        case "event":
+            String[] eventArgs = getEventArgs(commandArg);
+            return new EventCommand(new Event(eventArgs[0], parseDateArg(eventArgs[1])));
+        case "bye":
+            if (!commandArg.equals("")) {
+                throw new DukeException("bye must not have any argument.");
+            }
+            return new ByeCommand();
+        default:
+            throw new DukeException("Sorry, I don't understand what that means :(");
+        }
```
I like your use of `switch case` statement which makes your code more readable.

Ugholaf commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,60 @@
+        this.taskDescription = taskDescription;
+        this.isDone = false;
+    }
+
+    public void isDoneSetter(boolean isDone) {
+        this.isDone = isDone;
+    }
```
Missing Javadoc comment here.

Ugholaf commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+        }
+    }
+
+    /**
+     * Returns the n-th task in the TaskList.
+     *
+     * @param n Index of task to be retrieved.
+     * @return n-th task in TaskList.
+     * @throws DukeException If n - 1 is outside the bounds of TaskList's size.
+     */
+    public Task getTaskN(int n) throws DukeException {
```
Would it be better to name this method `getTaskByIndex` as N is confusing to readers what it refers to?

Rachel-AG replied:
That's true! I will rename the method in future commits

Ugholaf reviewed (commented):
Overall I like how you have structured your code and your naming conventions. The JavaDocs are also written very well! Good coding convention and code quality which makes the whole code very easy to read and understand!

Rachel-AG committed:
Fix checkstyle violations

Rachel-AG committed:
Add assert statements

Rachel-AG committed:
Improve code quality

Some of the methods such as the parser method and the load method are
quite long and have arrowhead style code which makes it difficult to
read and understand.

Long methods have been refactored to multiple other methods. Single
level of abstraction principle implemented when appropriate.

This allow for better readibility.

There are also some changes in method names to make it less ambiguous.

Rachel-AG committed:
Merge pull request #4 from Rachel-AG/branch-A-Assertions

Add assertions to code

Rachel-AG committed:
Add sorting functionality

Users might want to easily see the most or least urgent task to do.

A sort command is added to duke. Users can access this command by typing
"sort ascending" or "sort descending" to order the tasks in the list
from latest to earliest and vice versa.

Giving users the option to choose ascending and descending allow for
more flexibility and convenience especially when the list is long.

Rachel-AG committed:
Merge pull request #6 from Rachel-AG/branch-Sort

Add sorting functionality

Rachel-AG committed:
Merge branch 'master' into branch-A-CodeQuality

Rachel-AG committed:
Merge pull request #5 from Rachel-AG/branch-A-CodeQuality

Improve code quality

Rachel-AG committed:
Add more JUnit tests and fix sorting functionality

Rachel-AG committed:
Add user guide and title of app

Rachel-AG committed:
Update README.md

Rachel-AG committed:
Fix user guide format

Rachel-AG committed:
Merge branch 'master' of https://github.com/Rachel-AG/ip

Rachel-AG committed:
Modify build.gradle for cross-platform

Rachel-AG committed:
Modify app window to be fixed size

Rachel-AG committed:
Update README.md
